In [1]:
from sklearn import datasets

In [2]:
wine = datasets.load_wine()
wine

{'DESCR': '.. _wine_dataset:\n\nWine recognition dataset\n------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 178 (50 in each of three classes)\n    :Number of Attributes: 13 numeric, predictive attributes and the class\n    :Attribute Information:\n \t\t- Alcohol\n \t\t- Malic acid\n \t\t- Ash\n\t\t- Alcalinity of ash  \n \t\t- Magnesium\n\t\t- Total phenols\n \t\t- Flavanoids\n \t\t- Nonflavanoid phenols\n \t\t- Proanthocyanins\n\t\t- Color intensity\n \t\t- Hue\n \t\t- OD280/OD315 of diluted wines\n \t\t- Proline\n\n    - class:\n            - class_0\n            - class_1\n            - class_2\n\t\t\n    :Summary Statistics:\n    \n    ============================= ==== ===== ======= =====\n                                   Min   Max   Mean     SD\n    ============================= ==== ===== ======= =====\n    Alcohol:                      11.0  14.8    13.0   0.8\n    Malic Acid:                   0.74  5.80    2.34  1.12\n    Ash:           

In [3]:
wine.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [4]:
x_data = wine['data']
x_data.shape, type(x_data)

((178, 13), numpy.ndarray)

In [5]:
y_data = wine['target']
y_data

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [6]:
import pandas as pd

In [7]:
df_x_data = pd.DataFrame(x_data)
df_x_data, df_x_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       178 non-null    float64
 1   1       178 non-null    float64
 2   2       178 non-null    float64
 3   3       178 non-null    float64
 4   4       178 non-null    float64
 5   5       178 non-null    float64
 6   6       178 non-null    float64
 7   7       178 non-null    float64
 8   8       178 non-null    float64
 9   9       178 non-null    float64
 10  10      178 non-null    float64
 11  11      178 non-null    float64
 12  12      178 non-null    float64
dtypes: float64(13)
memory usage: 18.2 KB


(        0     1     2     3      4   ...    8      9     10    11      12
 0    14.23  1.71  2.43  15.6  127.0  ...  2.29   5.64  1.04  3.92  1065.0
 1    13.20  1.78  2.14  11.2  100.0  ...  1.28   4.38  1.05  3.40  1050.0
 2    13.16  2.36  2.67  18.6  101.0  ...  2.81   5.68  1.03  3.17  1185.0
 3    14.37  1.95  2.50  16.8  113.0  ...  2.18   7.80  0.86  3.45  1480.0
 4    13.24  2.59  2.87  21.0  118.0  ...  1.82   4.32  1.04  2.93   735.0
 ..     ...   ...   ...   ...    ...  ...   ...    ...   ...   ...     ...
 173  13.71  5.65  2.45  20.5   95.0  ...  1.06   7.70  0.64  1.74   740.0
 174  13.40  3.91  2.48  23.0  102.0  ...  1.41   7.30  0.70  1.56   750.0
 175  13.27  4.28  2.26  20.0  120.0  ...  1.35  10.20  0.59  1.56   835.0
 176  13.17  2.59  2.37  20.0  120.0  ...  1.46   9.30  0.60  1.62   840.0
 177  14.13  4.10  2.74  24.5   96.0  ...  1.35   9.20  0.61  1.60   560.0
 
 [178 rows x 13 columns], None)

In [8]:
df_y_data = pd.DataFrame(y_data)
df_y_data, df_y_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   0       178 non-null    int64
dtypes: int64(1)
memory usage: 1.5 KB


(     0
 0    0
 1    0
 2    0
 3    0
 4    0
 ..  ..
 173  2
 174  2
 175  2
 176  2
 177  2
 
 [178 rows x 1 columns], None)

In [9]:
import sqlite3

In [10]:
connect = sqlite3.connect('./db.sqlite3')

In [11]:
df_x_data.to_sql('wine_x', connect, if_exists='append', index=False)

In [12]:
df_y_data.to_sql('wine_y', connect, if_exists='append', index=False)

In [13]:
load_wine_x = pd.read_sql_query('select * from wine_x', connect)
load_wine_x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       178 non-null    float64
 1   1       178 non-null    float64
 2   2       178 non-null    float64
 3   3       178 non-null    float64
 4   4       178 non-null    float64
 5   5       178 non-null    float64
 6   6       178 non-null    float64
 7   7       178 non-null    float64
 8   8       178 non-null    float64
 9   9       178 non-null    float64
 10  10      178 non-null    float64
 11  11      178 non-null    float64
 12  12      178 non-null    float64
dtypes: float64(13)
memory usage: 18.2 KB


In [14]:
load_wine_y = pd.read_sql_query('select * from wine_y', connect)
load_wine_y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   0       178 non-null    int64
dtypes: int64(1)
memory usage: 1.5 KB


In [15]:
x_data = load_wine_x.to_numpy()
x_data.shape

(178, 13)

In [16]:
import numpy as np

In [17]:
y_data = load_wine_y.to_numpy()
np.unique(y_data)

array([0, 1, 2])

# 교육단계

In [50]:
import tensorflow as tf

In [51]:
model = tf.keras.Sequential()

In [52]:
model.add(tf.keras.Input(shape=13,))

model.add(tf.keras.layers.Dense(64, activation='relu'))
model.add(tf.keras.layers.Dense(36, activation='relu'))
model.add(tf.keras.layers.Dense(3, activation='softmax'))

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['acc'])

In [53]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 64)                896       
_________________________________________________________________
dense_11 (Dense)             (None, 36)                2340      
_________________________________________________________________
dense_12 (Dense)             (None, 3)                 111       
Total params: 3,347
Trainable params: 3,347
Non-trainable params: 0
_________________________________________________________________


In [54]:
model.fit(x_data, y_data, epochs=500, validation_split=0.3)

Epoch 1/500
4/4 [==============================] - 1s 49ms/step - loss: 88.0895 - acc: 0.5242 - val_loss: 44.9187 - val_acc: 0.1111
Epoch 2/500
4/4 [==============================] - 0s 10ms/step - loss: 27.8121 - acc: 0.5323 - val_loss: 17.4869 - val_acc: 0.0000e+00
Epoch 3/500
4/4 [==============================] - 0s 9ms/step - loss: 11.2875 - acc: 0.4758 - val_loss: 28.0420 - val_acc: 0.0000e+00
Epoch 4/500
4/4 [==============================] - 0s 8ms/step - loss: 19.7956 - acc: 0.4758 - val_loss: 29.3293 - val_acc: 0.0000e+00
Epoch 5/500
4/4 [==============================] - 0s 9ms/step - loss: 16.6995 - acc: 0.4758 - val_loss: 25.2862 - val_acc: 0.0000e+00
Epoch 6/500
4/4 [==============================] - 0s 10ms/step - loss: 6.7616 - acc: 0.5081 - val_loss: 19.3403 - val_acc: 0.0926
Epoch 7/500
4/4 [==============================] - 0s 13ms/step - loss: 3.5058 - acc: 0.6371 - val_loss: 26.2689 - val_acc: 0.1111
Epoch 8/500
4/4 [==============================] - 0s 10ms/step -

# evaluation

In [55]:
model.evaluate(x_data, y_data)

6/6 [==============================] - 0s 2ms/step - loss: 7.0011 - acc: 0.7191


[7.001112461090088, 0.7191011309623718]

* dense : 64, dense1 : 24, epochs : 50 --> loss: 8.3743 - acc: 0.6629 [8.374309539794922, 0.6629213690757751]
* dense : 64, dense1 : 24, epochs : 100 --> loss: 7.7834 - acc: 0.6910 [7.783389091491699, 0.6910112500190735]   val_loss: 25.3287 - val_acc: 0.1111
* dense : 48, dense1 : 24, epochs : 500 --> loss: 22.7629 - acc: 0.7079   [22.76287841796875, 0.7078651785850525]   acc: 0.9677 - val_loss: 74.8429 - val_acc: 0.1111
* dense : 64, dense1 : 48, epochs : 500 --> loss: 11.5153 - acc: 0.7191   [11.515317916870117, 0.7191011309623718]   acc: 0.9516 - val_loss: 37.7901 - val_acc: 0.1111
* dense : 64, dense1 : 48, epochs : 1000 --> loss: 13.2306 - acc: 0.7135   [13.230615615844727, 0.7134831547737122]    acc: 0.9839 - val_loss: 43.4997 - val_acc: 0.1111
* dense : 64, dense1 : 36, epochs : 500 --> loss: 7.0011 - acc: 0.7191  [7.001112461090088, 0.7191011309623718]   acc: 0.9677 - val_loss: 22.9039 - val_acc: 0.1111

* dense : 64, dense1 : 24, epochs : 500 --> loss: 6.9521 - acc: 0.7191  [6.952107906341553, 0.7191011309623718]  val_loss: 22.7861 - val_acc: 0.1111
* dense : 64, dense1 : 24, dense2 : 64 epochs : 100 --> loss: 2.0083 - acc: 0.7022   [2.008345365524292, 0.7022472023963928]     acc: 0.9839 - val_loss: 6.3024 - val_acc: 0.
* dense : 64, dense1 : 24, dense2 : 64 dense3 : 24 epochs : 100 --> loss: 1.6846 - acc: 0.5506   [1.6846214532852173, 0.550561785697937]    acc: 0.7500 - val_loss: 4.4220 - val_acc: 0.0926

In [56]:
x_data[38]

array([1.307e+01, 1.500e+00, 2.100e+00, 1.550e+01, 9.800e+01, 2.400e+00,
       2.640e+00, 2.800e-01, 1.370e+00, 3.700e+00, 1.180e+00, 2.690e+00,
       1.020e+03])

In [57]:
y_data[38]

array([0])

In [58]:
pred = model.predict([[1.307e+01, 1.500e+00, 2.100e+00, 1.550e+01, 9.800e+01, 2.400e+00,
       2.640e+00, 2.800e-01, 1.370e+00, 3.700e+00, 1.180e+00, 2.690e+00,
       1.020e+03]])
pred

array([[9.9747258e-01, 2.5273964e-03, 5.4945900e-16]], dtype=float32)

In [59]:
np.argmax(pred)

0